From: Michael Matz <matz@suse.de>
Subject: fixup inline asm for kexec
References: 147851

Fix a number of errors in the kexec asm for i386, x86-64 and ppc

diff -urp linux-2.6.15.suse/arch/i386/kernel/machine_kexec.c linux-2.6.15/arch/i386/kernel/machine_kexec.c
--- linux-2.6.15.suse/arch/i386/kernel/machine_kexec.c	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6.15/arch/i386/kernel/machine_kexec.c	2006-02-08 04:19:13.000000000 +0100
@@ -116,13 +116,13 @@ static void load_segments(void)
 	__asm__ __volatile__ (
 		"\tljmp $"STR(__KERNEL_CS)",$1f\n"
 		"\t1:\n"
-		"\tmovl $"STR(__KERNEL_DS)",%eax\n"
-		"\tmovl %eax,%ds\n"
-		"\tmovl %eax,%es\n"
-		"\tmovl %eax,%fs\n"
-		"\tmovl %eax,%gs\n"
-		"\tmovl %eax,%ss\n"
-		);
+		"\tmovl $"STR(__KERNEL_DS)",%%eax\n"
+		"\tmovl %%eax,%%ds\n"
+		"\tmovl %%eax,%%es\n"
+		"\tmovl %%eax,%%fs\n"
+		"\tmovl %%eax,%%gs\n"
+		"\tmovl %%eax,%%ss\n"
+		::: "eax", "memory");
 #undef STR
 #undef __STR
 }
diff -urp linux-2.6.15.suse/arch/x86_64/kernel/machine_kexec.c linux-2.6.15/arch/x86_64/kernel/machine_kexec.c
--- linux-2.6.15.suse/arch/x86_64/kernel/machine_kexec.c	2006-01-03 04:21:10.000000000 +0100
+++ linux-2.6.15/arch/x86_64/kernel/machine_kexec.c	2006-02-08 04:21:13.000000000 +0100
@@ -140,7 +140,7 @@ static void load_segments(void)
 		"\tmovl %0,%%ss\n"
 		"\tmovl %0,%%fs\n"
 		"\tmovl %0,%%gs\n"
-		: : "a" (__KERNEL_DS)
+		: : "a" (__KERNEL_DS) : "memory"
 		);
 }
 
diff -urp linux-2.6.15.suse/include/asm-powerpc/kexec.h linux-2.6.15/include/asm-powerpc/kexec.h
--- linux-2.6.15.suse/include/asm-powerpc/kexec.h	2006-02-08 04:10:24.000000000 +0100
+++ linux-2.6.15/include/asm-powerpc/kexec.h	2006-02-08 04:31:15.000000000 +0100
@@ -93,7 +93,8 @@ static inline void crash_setup_regs(stru
 			"mfxer  %0\n"
 			"std    %0, 296(%2)\n"
 			: "=&r" (tmp1), "=&r" (tmp2)
-			: "b" (newregs));
+			: "b" (newregs)
+			: "memory");
 	}
 }
 #else
